File: 2022/10/solution.py
import os
file1 = open(os.path.join(os.path.dirname(os.path.abspath(__file__)), 'input.txt'),'r')
lines = file1.readlines()

register = 1
cycle = 0
cycles = [20, 60, 100, 140, 180, 220]
strength = 0
screen = [[' '] * 40 for _ in range(6)]

def tick():
  global cycle, strength, screen
  if (register - 1) <= (cycle % 40) <= (register + 1):
    screen[cycle // 40][cycle % 40] = '#'
  cycle += 1
  if cycle in cycles:
    strength += cycle * register

for line in lines:
  if line[0] == 'n':
    tick()
  else:
    parse = line.split(' ')
    tick()
    tick()
    register += int(parse[1])

print(strength)

for line in screen:
  print(''.join(line))

File: 2022/10/solution.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strconv"
	"strings"

	"github.com/weitecklee/adventofcode/utils"
)

func main() {
	_, filename, _, _ := runtime.Caller(0)
	dirname := filepath.Dir(filename)
	inputFilePath := filepath.Join(dirname, "input.txt")
	data, err := os.ReadFile(inputFilePath)
	if err != nil {
		panic(err)
	}
	instructions := parseInput(strings.Split(string(data), "\n"))
	fmt.Println(solve(instructions))
}

type Instruction struct {
	ins string
	val int
}

func parseInput(data []string) []*Instruction {
	instructions := make([]*Instruction, len(data))
	for i, line := range data {
		if line == "noop" {
			instructions[i] = &Instruction{"noop", 0}
		} else {
			parts := strings.Split(line, " ")
			n, err := strconv.Atoi(parts[1])
			if err != nil {
				panic(err)
			}
			instructions[i] = &Instruction{parts[0], n}
		}
	}
	return instructions
}

func solve(instructions []*Instruction) (int, string) {
	registerX := 1
	cycle := 0
	cycleHistory := make(map[int]int)
	cycleHistory[cycle] = registerX
	for _, instruction := range instructions {
		cycle++
		cycleHistory[cycle] = registerX
		if instruction.ins == "addx" {
			cycle++
			registerX += instruction.val
			cycleHistory[cycle] = registerX
		}
	}
	res := 0
	for i := 20; i <= 220; i += 40 {
		res += i * cycleHistory[i-1]
	}
	var sb strings.Builder
	for cycle := 0; cycle < len(cycleHistory); cycle++ {
		if cycle%40 == 0 {
			sb.WriteByte('\n')
		}
		if utils.AbsInt(cycleHistory[cycle]-(cycle%40)) <= 1 {
			sb.WriteRune('█')
		} else {
			sb.WriteByte(' ')
		}
	}
	return res, sb.String()
}
